apis/core/v1beta1/featureflag_webhook.go

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"encoding/json"
+	"fmt"
+
+	_ "embed"
+
+	"github.com/xeipuuv/gojsonschema"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+//go:embed schema/targeting.json
+var TargetingSchema string
+
+//go:embed schema/flags.json
+var FlagsScheme string
+
+// log is for logging in this package.
+var featureflaglog = logf.Log.WithName("featureflag-resource")
+
+func (r *FeatureFlag) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+//+kubebuilder:webhook:path=/validate-core-openfeature-dev-v1beta1-featureflag,mutating=false,failurePolicy=fail,sideEffects=None,groups=core.openfeature.dev,resources=featureflags,verbs=create;update,versions=v1beta1,name=vfeatureflag.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Validator = &FeatureFlag{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *FeatureFlag) ValidateCreate() error {
+	featureflaglog.Info("validate create", "name", r.Name)
+
+	if err := validateFeatureFlagFlags(r.Spec.FlagSpec.Flags); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *FeatureFlag) ValidateUpdate(old runtime.Object) error {
+	featureflaglog.Info("validate update", "name", r.Name)
+
+	if err := validateFeatureFlagFlags(r.Spec.FlagSpec.Flags); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *FeatureFlag) ValidateDelete() error {
+	featureflaglog.Info("validate delete", "name", r.Name)
+
+	return nil
+}
+
+func validateFeatureFlagFlags(flags Flags) error {
+	b, err := json.Marshal(flags)
+	if err != nil {
+		return err
+	}
+
+	documentLoader := gojsonschema.NewStringLoader(string(b))
+	schemaLoader := gojsonschema.NewSchemaLoader()
+	schemaLoader.AddSchemas(gojsonschema.NewStringLoader(TargetingSchema))
+	compiledSchema, err := schemaLoader.Compile(gojsonschema.NewStringLoader(FlagsScheme))
```

Comment:
minor - I know we don't do this in flagd, but I think compiled schema is reusable. If so we should reuse it.

Reply:
I was thinking about this as well, but the only solution might be to to store it to a global var, which I would like to avoid. Webhook is a one-time execution on a single CR create/update. Not sure how it works behind the scenes, but we may get some weird errors when starting to store things in global vars.

Or did you had another solution in mind?

Reply:
The type/schema registration happens here [1], and here we simply create a struct. But yeah, I was also unable to find any details of the impact of having non API bound fields 

[1] - https://github.com/open-feature/open-feature-operator/blob/main/apis/core/v1beta1/featureflag_types.go#L86

Reply:
Already a singleton initialization is happening here. Hope this fits your ideas